Tools/clinic/clinic.py

```diff
@@ -5409,29 +5496,32 @@ def parse_converter(
                     "Annotations must be either a name, a function call, or a string."
                 )
 
-    def parse_deprecated_positional(self, thenceforth: str) -> None:
+    def parse_version(self, thenceforth: str) -> VersionTuple:
         assert isinstance(self.function, Function)
         fname = self.function.full_name
 
-        if self.keyword_only:
-            fail(f"Function {fname!r}: '* [from ...]' must come before '*'")
-        if self.deprecated_positional:
-            fail(f"Function {fname!r} uses '[from ...]' more than once.")
         try:
             major, minor = thenceforth.split(".")
-            self.deprecated_positional = int(major), int(minor)
+            return int(major), int(minor)
         except ValueError:
             fail(
-                f"Function {fname!r}: expected format '* [from major.minor]' "
+                f"Function {fname!r}: expected format '[from major.minor]' "
                 f"where 'major' and 'minor' are integers; got {thenceforth!r}"
             )
 
-    def parse_star(self, function: Function) -> None:
+    def parse_star(self, function: Function, version: VersionTuple | None) -> None:
         """Parse keyword-only parameter marker '*'."""
-        if self.keyword_only:
-            fail(f"Function {function.name!r} uses '*' more than once.")
-        self.deprecated_positional = None
-        self.keyword_only = True
+        if version is None:
+            if self.keyword_only:
+                fail(f"Function {function.name!r} uses '*' more than once.")
+            self.check_remaining_star()
+            self.keyword_only = True
+        else:
+            if self.keyword_only:
+                fail(f"Function {function.full_name!r}: '* [from ...]' must come before '*'")
+            if self.deprecated_positional:
+                fail(f"Function {function.full_name!r} uses '* [from ...]' more than once.")
```

Comment:
We should decide which variant to chose: `function.full_name` or `function.name`. IMO, `.name` should be sufficient. After I improved the accuracy of the line numbers, it now very easy to locate the offending line.